libvast/vast/arrow_extension_types.hpp

```diff
@@ -158,24 +173,350 @@ class pattern_extension_type : public arrow::ExtensionType {
+    return static_cast<T&>(x);
+  }
+
+  template <class T, int Index>
+  static const T*
+  get_if(const arrow::DataType* x, sum_type_token<T, Index> token) {
+    if (x && is(*x, token))
+      return &get(*x, token);
+    return nullptr;
+  }
+
+  template <class T, int Index>
+  static T* get_if(arrow::DataType* x, sum_type_token<T, Index> token) {
+    if (x && is(*x, token))
+      return &get(*x, token);
+    return nullptr;
+  }
+
+  template <class Result, class Visitor, class... Args>
+  static auto apply(const arrow::DataType& x, Visitor&& v, Args&&... xs)
+    -> Result {
+    // A dispatch table that maps variant type index to dispatch function for
+    // the concrete type.
+    static constexpr auto table = []<class... Ts, int... Indices>(
+      detail::type_list<Ts...>,
+      std::integer_sequence<int, Indices...>) noexcept {
+      return std::array{
+        // decay to function pointer
+        +[](const arrow::DataType& x, Visitor&& v, Args&&... xs) -> Result {
+          auto xs_as_tuple = std::forward_as_tuple(xs...);
+          auto indices = detail::get_indices(xs_as_tuple);
+          return detail::apply_args_suffxied(
+            std::forward<decltype(v)>(v), std::move(indices), xs_as_tuple,
+            get(x, sum_type_token<Ts, Indices>{}));
+        }...};
+    }
+    (types{}, std::make_integer_sequence<int, detail::tl_size<types>::value>());
+    const auto dispatch
+      = table[sum_type_access<arrow::Array>::index_from_type(x)];
+    VAST_ASSERT(dispatch);
+    return dispatch(x, std::forward<Visitor>(v), std::forward<Args>(xs)...);
+  }
+};
+
+namespace detail {
+template <class Types>
+struct tl_map_shared_ptr;
+template <class... Ts>
+struct tl_map_shared_ptr<detail::type_list<Ts...>> {
+  using type = detail::type_list<std::shared_ptr<Ts>...>;
+};
+
+} // namespace detail
+
+template <>
+struct sum_type_access<std::shared_ptr<arrow::Array>> final {
+  using types = typename detail::tl_map_shared_ptr<
+    typename sum_type_access<arrow::Array>::types>::type;
+  using type0 = detail::tl_head_t<types>;
+  static constexpr bool specialized = true;
+
+  template <class T, int Index>
+  static bool
+  is(const std::shared_ptr<arrow::Array>& arr, sum_type_token<T, Index>) {
+    VAST_ASSERT(arr);
+    if (arr->type_id() != T::element_type::TypeClass::type_id)
+      return false;
+    if constexpr (sum_type_access<arrow::Array>::is_extension_array<
+                    typename T::element_type>::value)
+      return static_cast<const arrow::ExtensionType&>(*arr->type())
+               .extension_name()
+             == T::element_type::TypeClass::vast_id;
+    return true;
+  }
+
+  template <class T, int Index>
+  static T
+  get(const std::shared_ptr<arrow::Array>& x, sum_type_token<T, Index>) {
+    return std::static_pointer_cast<typename T::element_type>(x);
+  }
+
+  template <class T, int Index>
+  static std::optional<T> get_if(const std::shared_ptr<arrow::Array>* x,
```

Comment:
I was wondering: Can't this just return a `T`, because `shared_ptr` already is nullable?

The same also applies for the `shared_ptr<DataType>` version below.

Reply:
We tried to do that but CAF disagrees, it insists on `get_if` to return a pointer to `shared_ptr`.